Prelecture HW

# 1.

In [43]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [44]:
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

# 2.

In [45]:
rows = df.shape[0]
columns = df.shape[1]
print(f"The number of rows in the villagers dataset is {rows} and the amount of columns is {columns}")

The number of rows in the villagers dataset is 391 and the amount of columns is 11


## Definition of Observations: 
### A data point corresponding to one instance in a dataset.

## Definition of Variables: 
### Characteristics or attributes of a data point containing the data which to be analyzed (i.e. the data/information about the observation).

# 3. 

In [46]:
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [47]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [48]:
print(df.columns)
    

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


In [49]:
df["gender"].value_counts()

gender
male      204
female    187
Name: count, dtype: int64

In [50]:
for column_name in df.columns:
    print("Column:", column_name)
    print(df[column_name].value_counts(dropna=False).sort_values(ascending=False))
    print("\n")

Column: row_n
row_n
2      1
117    1
4      1
6      1
7      1
      ..
475    1
477    1
480    1
481    1
483    1
Name: count, Length: 391, dtype: int64


Column: id
id
admiral     1
cranston    1
agnes       1
al          1
alfonso     1
           ..
winnie      1
wolfgang    1
yuka        1
zell        1
zucker      1
Name: count, Length: 391, dtype: int64


Column: name
name
Admiral     1
Cranston    1
Agnes       1
Al          1
Alfonso     1
           ..
Winnie      1
Wolfgang    1
Yuka        1
Zell        1
Zucker      1
Name: count, Length: 391, dtype: int64


Column: gender
gender
male      204
female    187
Name: count, dtype: int64


Column: species
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
deer         10
ostrich      10
gorilla       9
chicken     

In [51]:
df["species"].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [52]:
#check gender disparities in data
print("\n\nmale\n")
print(df[df["gender"]=="male"]["species"].value_counts(normalize=True))
print("\n\nfemale\n")
df[df["gender"]=="female"]["species"].value_counts(normalize=True)



male

species
frog         0.063725
bird         0.044118
bear         0.044118
cat          0.044118
dog          0.044118
rabbit       0.039216
horse        0.039216
penguin      0.039216
pig          0.039216
cub          0.039216
eagle        0.034314
lion         0.034314
duck         0.034314
mouse        0.034314
gorilla      0.034314
wolf         0.029412
bull         0.029412
elephant     0.029412
deer         0.029412
squirrel     0.024510
chicken      0.024510
alligator    0.024510
hamster      0.024510
tiger        0.019608
hippo        0.019608
goat         0.019608
monkey       0.019608
koala        0.019608
rhino        0.014706
ostrich      0.014706
sheep        0.014706
anteater     0.014706
octopus      0.009804
kangaroo     0.009804
Name: proportion, dtype: float64


female



species
cat          0.074866
squirrel     0.069519
rabbit       0.064171
sheep        0.053476
duck         0.053476
cub          0.042781
mouse        0.042781
ostrich      0.037433
pig          0.037433
dog          0.037433
horse        0.037433
kangaroo     0.032086
bear         0.032086
koala        0.026738
penguin      0.026738
elephant     0.026738
frog         0.026738
wolf         0.026738
deer         0.021390
cow          0.021390
bird         0.021390
monkey       0.021390
chicken      0.021390
goat         0.021390
anteater     0.021390
hamster      0.016043
rhino        0.016043
hippo        0.016043
tiger        0.016043
alligator    0.010695
eagle        0.010695
gorilla      0.010695
octopus      0.005348
Name: proportion, dtype: float64

# 4.

In [53]:
df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")
df2.shape

(891, 15)

In [54]:
print(f"Columns: {len(df2.columns)}")
print(f"Columns: {df2.shape[1]}")
df2.head()

Columns: 15
Columns: 15


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [55]:
print(f"Columns: {len(df2.describe().columns)}")
print(f"Columns: {((df2.dtypes== 'int64')|(df2.dtypes=='float64')).sum()}")
df2.describe()

Columns: 6
Columns: 6


survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [59]:
df2["age"].value_counts(dropna=False)

age
NaN      177
24.00     30
22.00     27
18.00     26
28.00     25
        ... 
36.50      1
55.50      1
0.92       1
23.50      1
74.00      1
Name: count, Length: 89, dtype: int64

In [60]:
print(f"total rows: {df2.shape[0]}")
print(f"total observations: {len(df2['age'])}")
print(f"non-null observations: {len(df2[df2['age'].isna()==False])}")

total rows: 891
total observations: 891
non-null observations: 714


In [61]:
df2.dtypes

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object

#### As seen from the above cells, there are a portion of columns which contain non integer or non float values and therefore the columns are not included in the column count for df2.describe() (which has 6 instead of 15). Columns such as 'age' have less than 891 values (the number of observations) due to the presence of 177 NaN entries (in the case of 'age'). This accounts for the discrepancy in the size of the dataset between that which is given by the describe method and shape attribute, as well as the discrepancy in the count of certain columns which from the describe method does not mirror the amount of observations as represented by the amount of rows in the shape attribute.

# 5.

## Definition of an Attribute: 
### An attribute is a piece of information about the object to which it is tied to. For example, df.shape stores a tuple which contains the dimensions of the dataset, or df.dtypes which returns a series storing the object type used for the attributes of each column.

## Definition of a Method: 
### A method is a function tied to the object which when called, allows one to perform a certain action.

Post Lecture HW

# 6.

#### The df.describe method computes the properties 'count', 'mean', etc. for every integer or float-valued column. 
count: count provides the amount of observations or column values used in computing the other properties for that specific column.

mean: mean computes the average value of the value in each observation for the given column

min: min returns the smallest value that appears in the column across all observations

std: the sample standard deviation represents the degree to which column values deviate from the norm. A too high std indicates the mean is not very representative of the full range of observation values for that column.

50%: 50% represents the median value of all values in the column, i.e. the value such that 50% of observations have values above this median, and 50% percent have a value below.

25%: In a similar fashion to 50%, 25% represents the value at which 25% of observations have values below the median, and 75% have values above the median

75%: 75% represents the value at which 75% of observations have values below the median, and 25% have values above the median

# 7.

##### 1. If the amount of observations that have non-null values in 'col' far exceeds the amount that do not, or the amount of observations with non-null values are independently statistically significant, df.dropna() would be preferred over del df['col'].

##### 2. If the amount of non-null values in the column do not provide statistically significant results, it would probably be preferred to drop the column entirely and thus retain the other column values for every observation (this need is itsefl yet another reason)

 ##### 3. If df['col'] has an amount of null values so large that removing all observations with non-null values in that column would harm the statistical significance of the dataset, it is recommended to first use del df['col'] and thus retain a lot more information when applying df.dropna() (which will now only get rid of the occasional anomalous data points with other missing values)

In [62]:
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [64]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [67]:
df['id'].value_counts(dropna=False)

id
admiral    1
muffy      1
paula      1
patty      1
pate       1
          ..
elvis      1
eloise     1
elmer      1
ellie      1
zucker     1
Name: count, Length: 391, dtype: int64

In [68]:
df["song"].value_counts(dropna=False)

song
NaN             11
K.K. Country    10
Forest Life      9
K.K. Lament      7
K.K. Ragtime     7
                ..
Aloha K.K.       2
Drivin'          1
Senor K.K.       1
K.K.  Bazaar     1
K.K. D&B         1
Name: count, Length: 93, dtype: int64

In [65]:
df.shape

(391, 11)

##### I shall first apply df.dropna() and get rid of the rows, as then I only lose 12 out of 391 observations. I will not remove the entire column if it contains missing values (as is the case in 'id' and 'song') because majority of the values are not null and therefore still useful.

In [69]:
df.dropna(inplace=True)

In [70]:
df['id'].value_counts(dropna=False)

id
admiral     1
olive       1
patty       1
pate        1
pashmina    1
           ..
eloise      1
elmer       1
ellie       1
elise       1
zucker      1
Name: count, Length: 379, dtype: int64

In [71]:
df["song"].value_counts(dropna=False)

song
K.K. Country     10
Forest Life       9
K.K. Soul         7
K.K. Ragtime      7
K.K. Lullaby      7
                 ..
K.K. Marathon     2
Drivin'           1
Senor K.K.        1
K.K.  Bazaar      1
K.K. D&B          1
Name: count, Length: 92, dtype: int64

In [72]:
df.shape

(379, 11)

In [73]:
df.isna().sum()

row_n          0
id             0
name           0
gender         0
species        0
birthday       0
personality    0
song           0
phrase         0
full_id        0
url            0
dtype: int64

# 8.

#### 1. The cell below sorts all the titanic dataset observations (of people) into two groups based on their sex, and then seperately selects the survived column in each of them and computes all values returned in the describe method.

In [77]:
df2.groupby('sex')['survived'].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

#### 2. The combination of count and groupby can help detect biases in the datasampling due to skews in selection of datapoint representtetive of general cases. Count alone merely helps ascertain how many values of a given column are null (when the value is compared to the amount of rows).

#### 3. For troubleshooting specific problems unique to my coding session, such as the ones in A-G, an AI is far more helpful than using a search engine for problem solving. Perhaps this lies in the very nature of Artificial Intelligence, which is meant to be able to work on data it was not neccessarily trained on, and the probability that you will find the exact same problem online is highly unlikley. More likely, you will end up spending a considerable amount of more time until you realize your own error, or until your stack overflow question is answered. Even if the EXACT same scenario IS present on a webpage, the chances that the search engine will provide you the right link is also low because formatting of code in a search engine can vary.

# 9. 

#### Mostly

# Log/Notes

## Assignment includes information garnered from earlier experience and pandas documentation, as well as AI.

### ChatGPT summarized log:
    Here's a summary of our chat log:

1. **2024-06-25 to 2024-09-09**: You have been working on various projects, including expanding image datasets, neural networks, machine learning, and financial tracking apps.
2. **Current Focus**: You are working on the Titanic dataset for summary analysis.
3. **Your Question**: You asked for an explanation of the code `df.groupby("col1")["col2"].describe()`.
4. **Explanation**: I explained that the code groups the dataset by `col1`, selects `col2` within each group, and then provides descriptive statistics for `col2` in each group (e.g., count, mean, standard deviation, etc.).

Let me know if you need more details!


### NotebookLM summarized log:
    ## Chat Log Summary: Exploring Data Manipulation and Statistics with Pandas

Our conversation explored various data manipulation and statistical concepts using the pandas library in Python. The sources you provided, which include excerpts from course materials for STA130, offer a structured outline of topics and serve as a valuable reference for this summary.

We started by addressing your query about counting species by gender. I suggested using the `df.groupby()` method with `value_counts()`, explaining how to group by 'gender' and count 'species' within those groups. This approach, while not explicitly mentioned in the sources, draws on the fundamental concepts of grouping and aggregation discussed in sources \.

Next, you asked about selecting rows where a specific column ('col1') equals 3. I provided the code `df_filtered = df[df['col1'] == 3]`, explaining the use of boolean indexing and filtering with a boolean mask. This response aligns with the concepts of subsetting via conditionals and boolean selection from sources \. 

You then wanted to select a DataFrame with certain data types. I outlined a three-step process:

1.  Identify the desired data types (e.g., `int64`, `float64`), referencing common pandas data types from source \.
2.  Create a boolean mask using comparison operators (`==`, `!=`) and logical operators (`&`, `|`), aligning with sources \.
3.  Apply the mask to the DataFrame to select columns, drawing on boolean indexing from sources \.

This response combines insights from various sources to provide a comprehensive approach to your query.

We then moved on to troubleshooting a `TypeError` in your code snippet: `print(f"Columns: {(df2.dtypes== ('int64'|'float64')).sum()}")`. I identified the issue—attempting to use the logical operator "or" (`|`) with strings instead of booleans. The solution involved separating the data type comparisons and then using the logical "or" on the resulting boolean values. This debugging process emphasized the importance of understanding data types and operator behavior, as highlighted in sources \.

Finally, you requested a good definition of standard deviation. My response provided a clear explanation of **sample standard deviation**, emphasizing its role in measuring data spread and its calculation using `df['column'].std()` or the formula: 

$$s = \sqrt{\frac{1}{n-1} \sum_{i=1}^{n} (x_i - \bar{x})^2}$$

This definition aligned with sources \, which explain how to calculate and interpret standard deviation as a statistical measure. 

Throughout our conversation, we addressed specific coding questions, clarified the use of pandas functions and attributes, and reinforced key statistical concepts. By integrating information from your provided sources and our interaction history, this summary aims to deepen your understanding of data manipulation and analysis with pandas.
